pattern/variable/SystemReference.java

```diff
@@ -18,17 +18,43 @@
 
 package grakn.core.pattern.variable;
 
-import graql.lang.common.GraqlToken;
 import graql.lang.pattern.variable.Reference;
 
+import java.util.Objects;
+
 public class SystemReference extends Reference.Name {
 
-    public SystemReference(final String name) {
-        super(name);
+    public static String PREFIX = "$/";
+
+    private final int id;
+    private final int hash;
+
+    public SystemReference(int id) {
+        super("sys" + id); // TODO: why are we we doing this? Why does this class have to extend Reference.Name?
+        this.id = id;
+        this.hash = Objects.hash(SystemReference.class, id);
+    }
+
+    public static SystemReference of(int id) {
+        return new SystemReference(id);
     }
 
     @Override
     public String syntax() {
-        return GraqlToken.Char.$SYS + name;
+        return PREFIX + name;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        SystemReference that = (SystemReference) o;
+        return this.id == that.id;
+    }
+
+    @Override
+    public int hashCode() {
+        return hash;
```

Comment:
This class did not have its own `equals()` and `hashCode()` function, which means it was borrowing the methods from its parent, `Reference.Name`, which had different logic that was invalid for this class.